plugin/db/mysql/binlog_stream_parser.go

```diff
@@ -0,0 +1,139 @@
+package mysql
+
+import (
+	"bufio"
+	"io"
+	"strings"
+
+	"github.com/pkg/errors"
+)
+
+// BinlogEventType is the enumeration of binlog event types.
+type BinlogEventType int
+
+const (
+	// DeleteRowsEvent is the binlog event for DELETE.
+	DeleteRowsEvent BinlogEventType = iota
+	// UpdateRowsEvent is the binlog event for UPDATE.
+	UpdateRowsEvent
+	// WriteRowsEvent is the binlog event for INSERT.
+	WriteRowsEvent
+	// QueryEvent is the binlog event for QUERY.
+	// The thread ID is parsed from QUERY events.
+	QueryEvent
+	// OtherEvent represents other types of event that are ignored.
+	OtherEvent
+)
+
+// BinlogEvent contains the raw string of a single binlog event from the mysqlbinlog output stream.
+type BinlogEvent struct {
+	Type   BinlogEventType
+	Header string
+	Body   string
+}
+
+// ParseBinlogStream splits the mysqlbinlog output stream to a list of transactions.
+// Each transaction is a list of events, starting with Query (BEGIN).
+func ParseBinlogStream(stream io.Reader) ([][]BinlogEvent, error) {
+	reader := bufio.NewReader(stream)
+	prevLineType := otherLine
+	eventType := OtherEvent
+	var eventHeader string
+	var bodyBuf strings.Builder
+	var txns [][]BinlogEvent
+	var txn []BinlogEvent
+	done := false
+	for !done {
```

Comment:
The code block below can be simplified A LOT. Think of 1) extra variables above, 2) less if-cases below.